internal/wasm/namespace.go

```diff
@@ -103,16 +126,16 @@ func (ns *Namespace) CloseWithExitCode(ctx context.Context, exitCode uint32) (er
 	ns.mux.Lock()
 	defer ns.mux.Unlock()
 	// Close modules in reverse initialization order.
-	for i := len(ns.moduleNamesList) - 1; i >= 0; i-- {
+	for node := ns.moduleNamesList; node != nil; node = node.next {
 		// If closing this module errs, proceed anyway to close the others.
-		if m, ok := ns.modules[ns.moduleNamesList[i]]; ok {
+		if m, ok := ns.modules[node.name]; ok {
 			if _, e := m.CallCtx.close(ctx, exitCode); e != nil && err == nil {
 				err = e // first error
 			}
 		}
 	}
 	ns.moduleNamesList = nil
-	ns.moduleNamesSet = map[string]struct{}{}
+	ns.moduleNamesSet = map[string]*nameListNode{}
```

Comment:
I don't think it would help much but there may be uses I'm not considering.

Something worth being aware of is that go wont shrink maps as elements are deleted. It may actually be beneficial in some cases to allocate a new map if the old one grew very large. It may be useful to track in the delete method.